### PR TITLE
Three new surface irrigation parameterizations

### PR DESCRIPTION
**Pull request by arjanna**
TYPE: new feature

KEYWORDS: three surface irrigation parameterizations, explicit namelist parameters

SOURCE: Arianna Valmassoi (University College Dublin/NCAR)

DESCRIPTION OF CHANGES: 
Three surface parameterizations for irrigation have been added. They consider different (cumulative) evaporative processes from irrigation water:
1. only from the surface
2. from the canopy/leaves interception
3. from the microphysics processes. 

The first two parameterizations are called in the surface driver and the irrigation water then is input to the land surface scheme. The third parameterization scheme is called in the microphysics driver. The MP driver modifies the rainwater mixing ratio in the lowest mass level, and the updated field is then passed to the chosen parameterization.

The parameterizations can be activated via the namelist switch `sf_surf_irr_scheme` (respectively option 1,2,3) and have explicit timing and water amount (in mm/day) setting from the namelist. The irrigation water amount is spatially modulated by the `IRRIGATION` mask derived from FAO dataset and given to geogrid. Both GEOGRID.TBL and METGRID.TBL need to be modified in WPS to interpolate the new irrigation mask field.

Physics schemes dependency: 
`sf_surf_irr_scheme = 1` works only with Noah LSM.
`sf_surf_irr_scheme = 2` and `sf_surf_irr_scheme = 3` both have no restrictions regarding physics scheme selections.

LIST OF MODIFIED FILES:
M       Registry/Registry.EM_COMMON
M       dyn_em/module_first_rk_step_part1.F
M       dyn_em/solve_em.F
M       dyn_em/start_em.F
M       phys/Makefile
A       phys/module_irrigation.F
M       phys/module_microphysics_driver.F
M       phys/module_physics_init.F
M       phys/module_sf_noahdrv.F
M       phys/module_sf_noahlsm.F
M       phys/module_surface_driver.F
M       run/README.namelist
M       share/module_check_a_mundo.F
M       test/em_real/examples.namelist

TESTS CONDUCTED: 
1. Model compiled with GNU and Intel for ARW core. NMM core compiles, no irrigation 
parameterizations.
2. Restart tested.
3. consistency check (diffwrf) with 1 and 4 processor for scheme 1,2,3 and no irrigation 
(using irr_ph=0 so no random numbers).
4. Auto jenkins test all PASS.

RELEASE NOTE: Adding three surface irrigation options that parameterize irrigation as a function of three evaporative processes that the water applied undergoes. The irrigation water can evaporate (i) from the surface, (ii) from the canopy and then drip on the surface, or (iii) from the atmospheric lowest mass-model layer and then follow the previous process. The schemes have an explicit treatment regarding the water amount, location, and timing.

### CODE REVIEW

**Comment by davegill** [comment]
@arjanna 
Arianna,
1. In the commit message, we use the "RELEASE NOTE" directly when we issue a release. Would you please add a couple of extra sentences to make this portion of the PR more stand-alone. You do not need as much detail as you will put into the "DESCRIPTION OF CHANGES".

2. Are there specific schemes that are able to use this new option and other schemes that are not? In the "DESCRIPTION OF CHANGES", please add information regarding which schemes are supported.

3. There are user-support files for namelists in the WRF repo. In both `test/em_real/examples.namelist` and `run/README.namelist`, add in information for your scheme. You'll see that one provides a working example, and the other provides a sentence or two of information about the option for the user.

4. For the "TESTS CONDUCTED" section, we are interested in knowing if this code works with some fundamental pieces of the WRF system: NMM core, Chemistry, does it give bit-wise identical results with nested restarts, does it work with digital filtering, does it work with adaptive time stepping, can you have a nest with only coarse-grid input.

5. If there are restrictions to the code (inconsistent physics choices, wrong input data, etc), those conditions should be caught in share/module_check_a_mundo.F and phys/module_physics_init.F.

**Comment by arjanna** [comment]
Dear Dave,
I modified point 1-5 and committed the new files. Is this correct?
Thank you!

File: Registry/Registry.EM_COMMON [hunk 1]
```diff
@@ -2895,6 +2907,11 @@ package   ssibscheme     sf_surface_physics==8       -             state:ssib_fm
 package   noahmpoptcrop2 opt_crop==2                 -             state:gecros_state
 
 package   noahmosaicscheme  sf_surface_mosaic==1         -        state:TSK_mosaic,QSFC_mosaic,TSLB_mosaic,SMOIS_mosaic,SH2O_mosaic,CANWAT_mosaic,SNOW_mosaic,SNOWH_mosaic,SNOWC_mosaic,ALBEDO_mosaic,ALBBCK_mosaic,EMISS_mosaic,EMBCK_mosaic,ZNT_mosaic,Z0_mosaic,HFX_mosaic,QFX_mosaic,LH_mosaic,GRDFLX_mosaic,SNOTIME_mosaic,TR_URB2D_mosaic,TB_URB2D_mosaic,TG_URB2D_mosaic,TC_URB2D_mosaic,TS_URB2D_mosaic,TS_RUL2D_mosaic,QC_URB2D_mosaic,UC_URB2D_mosaic,TRL_URB3D_mosaic,TBL_URB3D_mosaic,TGL_URB3D_mosaic,SH_URB2D_mosaic,LH_URB2D_mosaic,G_URB2D_mosaic,RN_URB2D_mosaic,mosaic_cat_index,landusef2,smcrel,RS_mosaic,LAI_mosaic
+# IRRIGATION
+package   drip           sf_surf_irr_scheme==2        -            -
```
Comment (davegill) [inline Registry/Registry.EM_COMMON]: @arjanna 
Would you put these in 1, 2, 3 order

File: dyn_em/solve_em.F [hunk 2]
```diff
@@ -3624,10 +3624,9 @@ SUBROUTINE solve_em ( grid , config_flags  &
 #endif
 
 
-     CALL microphysics_driver(                                            &
-      &         DT=dtm             ,DX=grid%dx              ,DY=grid%dy   &
+     CALL microphysics_driver(DT=dtm             ,DX=grid%dx              ,DY=grid%dy   &
```
Comment (davegill) [inline dyn_em/solve_em.F]: @arjanna 
You cleaned up some indenting, so fix the unnecessary mods to the dx and dy line

File: phys/Makefile [hunk 3]
```diff
@@ -189,7 +189,8 @@ MODULES = \
 	module_fddaobs_driver.o \
 	module_wind_fitch.o \
 	module_sf_lake.o \
-	module_diagnostics_driver.o
+	module_diagnostics_driver.o\
```
Comment (davegill) [inline phys/Makefile]: @arjanna 
Arianna,
Add a space before the backslash

File: phys/module_microphysics_driver.F [hunk 4]
```diff
@@ -721,6 +726,16 @@ smaw3d, &    !Normalized Galperin instability function for momentum [no units]
...
+! IRRIGATION
+
+   REAL, DIMENSION( ims:ime , jms:jme ), INTENT(IN):: IRRIGATION !ARI
+   REAL,  INTENT(IN)::  irr_daily_amount, xtime, gmt, julian_in
+   INTEGER, INTENT(IN ):: sf_surf_irr_scheme, irr_start_hour, irr_num_hours,irr_start_julianday,irr_freq,irr_ph
+   REAL, PARAMETER    :: PI_GRECO=3.14159
+   INTEGER  :: end_hour,a,b,xt24,irr_day,timing
+   REAL :: constants_irrigation,tloc,irr_start,phase
+   INTEGER, DIMENSION( ims:ime , jms:jme ), INTENT(INOUT) :: irr_rand_field
+   REAL, DIMENSION( ims:ime, kms:kme, jms:jme ) ::       dummy_qr
```
Comment (davegill) [inline phys/module_microphysics_driver.F]: @arjanna 
Arianna,
Usually, the "dummy" type arrays are tile-sized.

File: phys/module_physics_init.F [hunk 5]
```diff
@@ -2990,7 +2998,9 @@ SUBROUTINE bl_init(STEPBL,BLDT,DT,RUBLTEN,RVBLTEN,RTHBLTEN,  &
                      allowed_to_read ,                         &
                      ids,ide, jds,jde, kds,kde,                &
                      ims,ime, jms,jme, kms,kme,                &
-                     its,ite, jts,jte, kts,kte                 )
+                     its,ite, jts,jte, kts,kte,                &
```
Comment (davegill) [inline phys/module_physics_init.F]: @arjanna 
Arianna,
Traditionally, we add new arguments PRIOR to the dimensions.

File: phys/module_sf_noahdrv.F [hunk 6]
```diff
@@ -105,7 +105,7 @@ SUBROUTINE lsm(DZ8W,QV3D,P8W3D,T3D,TSK,                      &
                   dl_u_bep,sf_bep,vl_bep,sfcheadrt,INFXSRT, soldrain      &   !O multi-layer urban         
                  ,SDA_HFX, SDA_QFX, HFX_BOTH, QFX_BOTH, QNORM, fasdas     &   !fasdas
                  ,RC2,XLAI2                                               &
-                  )
+                 ,IRR_CHAN)
```
Comment (davegill) [inline phys/module_sf_noahdrv.F]: @arjanna 
Arianna,
Again, keeping with conventions, your added code would be 
```
   ,IRR_CHAN       &
```
Insert you code prior to the closing ")".

File: phys/module_surface_driver.F [hunk 7]
```diff
@@ -1480,6 +1501,12 @@ SUBROUTINE surface_driver(                                         &
          RAINBL(i,j) = RAINBL(i,j) + RAINCV(i,j) + RAINNCV(i,j)
          IF ( PRESENT( rainshv ))RAINBL(i,j) = RAINBL(i,j) + RAINSHV(i,j)
          RAINBL(i,j) = MAX (RAINBL(i,j), 0.0)
+         sf_surf_irr: SELECT CASE(sf_surf_irr_scheme)
+            CASE(DRIP)
+               CALL drip_irrigation(julian_in,IRRIGATION(i,j),sf_surf_irr_scheme, irr_daily_amount,irr_start_hour,irr_num_hours,irr_start_julianday,irr_freq,irr_ph,i,j,RAINBL(i,j),IRRIGATION_CHANNEL(i,j),gmt,xtime,dt,irr_rand_field(i,j))
```
Comment (davegill) [inline phys/module_surface_driver.F]: @arjanna 
Arianna,
Use continuation "&" characters for lines over the Fortran standard.

File: run/README.namelist [hunk 8]
```diff
@@ -691,6 +691,20 @@ Namelist variables for controlling the adaptive time step option:
  num_urban_nf                        = 1! (= 10 if BEM active)        number of grid levels in the floors (nf_u in BEM header)
  num_urban_nbui                      = 1! (= 15 if BEM active)        maximum number of types of buildings in an urban class (nbui_max in BEM header)
 
+ sf_surf_irr_scheme (max_dom)        = 0: no irrigation scheme
+                                     = 1: channel scheme (in Noah only)
+                                     = 2: drip scheme
+                                     = 3: sprinkler scheme
+ 
+ irr_daily_amount (max_dom)          = daily equivalent irrigation water amount in mm (e.g. 5.7)
+ irr_start_hour (max_dom)            = UTC start hour for irrigation (e.g. 5)
```
Comment (davegill) [inline run/README.namelist]: @arjanna 
Arianna,
For the "hour" options, please indicate if these are "real" or "integer" values.

**Comment by davegill** [comment]
@arjanna 
Arianna,
Here are a few more ...
1. Would you look over the "tweaked" pull request commit message. Please edit it for accuracy.
2. The mods to check_a_mundo look like you need to add in a test for your option. Right now, if you ask for Noah, the code will stop, right?
3. What is the situation with `test/em_real/wind-turbine-1.tbl `? Did you commit the file and then delete it?

**Comment by arjanna** [comment]
Hi Dave,
I modified the files accordingly to your comments and committed the changes. The check_a_mundo should be correct now.
> 
>     3. What is the situation with `test/em_real/wind-turbine-1.tbl `? Did you commit the file and then delete it?
The file was committed as changed by mistake. I did not use it/modify it, should I re-commit it?

**Comment by dudhia** [comment]
Arianna, the developers would like the name of your module changed to something more generic. Can we do that? In fact module_sf_irrigation might be better.

**Comment by arjanna** [comment]
> Arianna, the developers would like the name of your module changed to something more generic. Can we do that? In fact module_sf_irrigation might be better.

I did not use "sf" (so module_irrigation) otherwise it might be confusing for the option "sprinkler" that is called by the microphysics driver (so not just by the surface modules). I have built it before committing the changes.

**Comment by dudhia** [comment]
I am OK with leaving the name as module_irrigation.F instead of module_sf_irrigation.F. If others have an opinion, post it here.

**Comment by davegill** [comment]
@arjanna 
The PR lists 13 file changed, and github thinks there are 15. Please sort out the missing two files and add them to the PR.

**Comment by dudhia** [comment]
Somehow the turbine file always shows up as a difference which is spurious. The old named file has not been removed yet, so I think this accounts for the 2 extra ones.

**Comment by davegill** [comment]
@dudhia @weiwangncar 
Jimy,
We need to make a decision on the `wind-turbine-1.tbl` file. We have two of them: one in `run` and the other in `test/em_real`. They are different.

Screen Shot 2020-02-06 at 12 47 39 PM

In the top-level `Makefile`, we overwrite the version in the `test/em_real` directory with the version from the `run` directory.

1. We should get rid of two identically named, yet different, files.
2. We need to choose which one to keep.
3. If the version in `test/em_real` is the gold standard, we need to remove the `ln -sf` command in the top-level `Makefile`.
4. We need to treat `windturbines.txt` identically to how we treat `wind-turbine-1.tbl`.

This is an easy fix to do. We'll do it outside of this PR.

**Comment by dudhia** [comment]
Yes, that explains why I always see differences in my git. We need to fix
this. I am cc'ing Pedro,
Jimy

**Comment by dudhia** [comment]
I you change it in em_real, does a recompile overwrite it?

**Comment by davegill** [comment]
>I you change it in em_real, does a recompile overwrite it?

Yes. The command `./compile em_real` will overwrite the `wind-turbine-1.tbl` file that was originally in the `test_em/real` directory before the build.

**Comment by dudhia** [comment]
That is a behavior we don't want. Seems we need identical copies in em_real
and run and no linking.

**Comment by davegill** [comment]
@dudhia 
Jimy,
Once you approve this and merge it, then I'll put together a PR to address the replicated input file. We need to know from Pedro which one we keep (the one in run, or the one in test/em_real).

**Comment by davegill** [comment]
@dudhia @arjanna 
Folks,
Please review the unsolicited commit I made:
https://github.com/wrf-model/WRF/pull/980/commits/216a485c4a99504c90863344e18aecf15b8aa84c

File: share/module_check_a_mundo.F [hunk 9]
```diff
@@ -401,6 +401,18 @@ END FUNCTION bep_bem_nbui_max
...
+!-----------------------------------------------------------------------
+      DO i = 1, model_config_rec % max_dom
+         IF ( model_config_rec % sf_surface_physics(i) .EQ. LSMSCHEME .AND.  &
+             model_config_rec % sf_surf_irr_scheme(i) .EQ. CHANNEL ) THEN
+              wrf_err_message = '--- ERROR: irrigation Opt 1 works only with Noah-LSM'
+              CALL wrf_message ( wrf_err_message )
+         count_fatal_error = count_fatal_error + 1
+         END IF
+      ENDDO
+
```
Comment (davegill) [inline share/module_check_a_mundo.F]: @arjanna @dudhia 
Is this IF test backwards? The comments say we must have LSM with channel. The IF test fails with that combination.

**Comment by davegill** [comment]
@arjanna @dudhia @weiwangncar 
Folks,
This code breaks NMM.
The microphysics driver references `sprinkler`.
The surface driver references `drip` and `channel`.
The physics init has several irrigation fields in the argument list.

The two physics drivers could be ifdef'ed for EM_CORE. However, the physics init explicitly needs those arguments. Therefore, it seems that most of the irrigation info in the Registry.EM_COMMON needs to get into Registry.NMM.

**Comment by dudhia** [comment]
Can it be done with optional arguments and IF(PRESENT) checks?

**Comment by davegill** [comment]
@dudhia @arjanna 
>Can it be done with optional arguments and IF(PRESENT) checks?

Jimy,
Yes. Then we would need `OPTIONAL` arguments, and we need `IF(PRESENT(xxx))` tests. 

We make NMM work both ways. We need Laurie's approval when we change the `dyn_nmm/module_PHYSICS_CALLS.F` file, so the `OPTIONAL` arguments might be the easier way to go.

This needs to be pretty soon ...

**Comment by dudhia** [comment]
It would help Arianna if you could list the variables that need to be
optional.
Perhaps all the new ones in the phys_init and physics driver arguments.
I could also try to do this when I have time in the week,
Jimy

**Comment by davegill** [comment]
## THIS ANSWER IS RIGHT

@dudhia @arjanna @weiwangncar 
>Can it be done with optional arguments and IF(PRESENT) checks?

No. The problem is phy_init. We do not have the capability to have optional variables in the call to phy_init without significant work.

All of the irrigation initialization variables and namelist options must either be put into the NMM Registry to be passed into phy_init from start_nmm.F, or the physics initialization routine will need to include ifdefs for EM_CORE.

**Comment by dudhia** [comment]
Why does not making them optional in phy_init not work to avoid having to
do anything
with NMM?

**Comment by dudhia** [comment]
It seems adding these irrigation fields at the end of the phy_init call
instead of the beginning
would allow them to be optional. phy_init has many optional variables.
These need to be at the end because it doesn't have explicit argument
assignments,
Jimy

File: share/module_check_a_mundo.F (hunk 9 above)
Comment (arjanna) [reply share/module_check_a_mundo.F]: In previous calls (and in the Registry.EM_COMMON.var) "lsmscheme" is Noah. If it's wrong like this, which would be the correct? Thanks!

**Comment by davegill** [comment]
@dudhia 
>We do not have the capability to have optional variables in the call to phy_init without significant work.

Jimy,
In this ONE instance, yes, we could put arguments at the end of the list. However, if we are really going to make phy_init able to have optional arguments (not a bad idea given how difficult it is to make NMM support some of these newer options), then we need to introduce the same subroutine call syntax that we use in the physics drivers:
```
call sub ( var1=grid%var1, var2=grid%var2, etc) 
```

1. There are about 630 arguments that need to be handled. 
2. The calls from both dyn_em/start_em.F and dyn_nmm/start_nmm.F need to be modified. The grid% variables names differ in the two dy cores.
3. The use of OPTIONAL inside of the phy_init routine needs to be fixed. For example, most of the `IF(PRESENT(xxx))` cases are always there. Right now, those statements do no harm as they are mostly not used.
4.  How long do we expect to support NMM? Usually the answer is until HWRF is retired from service. Is this "optional argument phy_init" task worth doing if support for the NMM dy core is going away?

**Comment by dudhia** [comment]
phy_init already has many optional variables at the end, eg. mosaic LSM.

File: share/module_check_a_mundo.F (hunk 9 above)
Comment (davegill) [reply share/module_check_a_mundo.F]: @arjanna @dudhia 
This IF test says
```
if ( using NOAH and CHANNEL ) then
   FAIL
```
Is this correct?
I though the logic would be something like this:
```
if ( using NOAH and CHANNEL ) then
   everything is OK
else
   FAIL
```

Comment (arjanna) [reply share/module_check_a_mundo.F]: yes sorry, I noticed now the mistake. I fixed it and I am checking the output now. I will commit the change with the new push.

**Comment by dudhia** [comment]
Apart from moving the irrigation arguments to the end of the phy_init list and making them optional, the lsminit code probably needs ifdef EM_CORE around the irr_ph etc code to make sure NMM doesn't execute the irrigation code.

File: phys/module_sf_noahdrv.F [hunk 10]
```diff
@@ -1776,6 +1781,19 @@ SUBROUTINE LSMINIT(VEGFRA,SNOW,SNOWC,SNOWH,CANWAT,SMSTAV,    &
 ! <--GAC
 
    IF(.not.restart)THEN
+   IF (irr_ph.NE.0)THEN
```
Comment (dudhia) [inline phys/module_sf_noahdrv.F]: This is where we need #if (EM_CORE==)

Comment (arjanna) [reply phys/module_sf_noahdrv.F]: Ok. I am making the changes also on the phy_init argument list so they work only with EM_CORE==1

**Comment by dudhia** [comment]
> @dudhia @weiwangncar
> Jimy,
> We need to make a decision on the `wind-turbine-1.tbl` file. We have two of them: one in `run` and the other in `test/em_real`. They are different.
> 
> Screen Shot 2020-02-06 at 12 47 39 PM
> 
> In the top-level `Makefile`, we overwrite the version in the `test/em_real` directory with the version from the `run` directory.
> 
> 1. We should get rid of two identically named, yet different, files.
> 2. We need to choose which one to keep.
> 3. If the version in `test/em_real` is the gold standard, we need to remove the `ln -sf` command in the top-level `Makefile`.
> 4. We need to treat `windturbines.txt` identically to how we treat `wind-turbine-1.tbl`.
> 
> This is an easy fix to do. We'll do it outside of this PR.

@pedro-jm we have two different turbine tables in run and em_real. We have to decide which one to keep as we do not want two different ones.

File: phys/module_sf_noahdrv.F [hunk 11]
```diff
@@ -785,6 +789,11 @@ SUBROUTINE lsm(DZ8W,QV3D,P8W3D,T3D,TSK,                      &
 ! use mid-day albedo to determine net downward solar (no solar zenith angle correction)
         SOLNET=SOLDN*(1.-ALBEDO(I,J))
         PRCP=RAINBL(i,j)/DT
+        IF(IRR_CHAN(i,j).NE.0) THEN
```
Comment (dudhia) [inline phys/module_sf_noahdrv.F]: May need a PRESENT check for IRR_CHAN here and in lsm_mosaic.
NMM will run Noah too, so any irrigation things have to be blocked for it.

Comment (arjanna) [reply phys/module_sf_noahdrv.F]: corrected now

File: phys/module_irrigation_arianna.F [hunk 12]
```diff
@@ -0,0 +1,166 @@
+MODULE module_irrigation_arianna
```
Comment (dudhia) [inline phys/module_irrigation_arianna.F]: make sure you remove this old module

File: share/module_check_a_mundo.F (hunk 9 above)
Comment (davegill) [reply share/module_check_a_mundo.F]: Yep, good to go

**Comment by davegill** [comment]
@arjanna @kkeene44 
Arianna,
Would you please send a description to Kelly Keene (kkeene@ucar.edu) regarding this scheme for the WRF User's Guide. For an example, take a look at: https://www2.mmm.ucar.edu/wrf/users/docs/user_guide_v4/v4.1/users_guide_chap5.html. 

Perhaps something similar to the section on "Stochastic Parameterization Schemes", you could put in something regarding "Irrigation Schemes". This can be made available to us over the next few weeks. Kelly will let you know how to send information (such as sending plain text, edit the github page, etc.). This is not time-critical compared to the PR.

**Comment by davegill** [comment]
@dudhia 
Jimy,
If the regression tests are all OK, please indicate that in the PR. The NMM builds were the only thing in the way, right?

**Comment by dudhia** [comment]
I am waiting for Arianna to fix the PR with my mods then we can run the
automated tests again.
Two more things are needed.
1. 1 versus 4 procs with the scheme on.
2. New versus old Noah LSM with the scheme off.
Both hopefully will produce identical results.
Jimy

**Comment by davegill** [comment]
@weiwangncar @smileMchen @kkeene44 
Please review

**Comment by weiwangncar** [comment]
@arjanna @dudhia Can you verify that the code before and after this addition gives the same results with Noah? If it has been tested, please add it to the PR.

Also, the PR mentions something needed from WPS/geogrid, can you provide the specifics?

**Comment by weiwangncar** [comment]
@arjanna @dudhia Can we remove these two diffs from wind-turbine-1.tbl from this PR?

**Comment by dudhia** [comment]
We need to.

**Comment by weiwangncar** [comment]
@arjanna @dudhia Please add module_sf_noahlsm.F as a modified file.

**Comment by davegill** [comment]
@weiwangncar 
Wei,
1. I added module_sf_noahlsm.F as a modified file
2. The wind-turbine-1.tbl problem is removed
3. Also, I put the original QSS namelist.input back in place

**Comment by kkeene44** [comment]
@arjanna 
If you just want to send me the description(s) that should be added to the WRF Users' Guide, I can make those updates - whenever you are ready. Thanks!

**Comment by arjanna** [comment]
> @arjanna @dudhia Can you verify that the code before and after this addition gives the same results with Noah? If it has been tested, please add it to the PR.
> 
> Also, the PR mentions something needed from WPS/geogrid, can you provide the specifics?

The schemes need an irrigation mask (variable "irrigation"), which is an invariant field. The IRRIGATION field has to be interpolated (METGRID/GEOGRID.TBL) as the other fields. I have the binary that Mike (Barlage) shared with me (it covers: 00001-04320.00001-02160). I can share the files with you.

**Comment by davegill** [comment]
@mgduda @weiwangncar @dudhia @arjanna @barlage @kkeene44 
> The schemes need an irrigation mask (variable "irrigation"), which is an invariant field. The IRRIGATION field has to be interpolated (METGRID/GEOGRID.TBL) as the other fields. I have the binary that Mike (Barlage) shared with me (it covers: 00001-04320.00001-02160). I can share the files with you.

Folks,
We seem to now have a "to do" item for WPS for release v4.2. If we are agreeing to this at this fairly late date:
1. We need the data from Barlage
2. We need the data in the standard WRF download location
3. We need mods to geogrid tables

**Comment by arjanna** [comment]
>     3. We need mods to geogrid tables
I can open a pull request with the two mods for the tables. Otherwise, I can past them here.

**Comment by barlage** [comment]
I have place the data here:

/glade/scratch/barlage/geogrid/geog

I suggest the nested directory structure since in our irrigation work we have multiple datasets and it is cleaner to have them collocated in an irrigation/ directory. My GEOGRID.TBL has the following entry:


===============================
name=IRRIGATION
        priority=1
        optional=yes
        dest_type=continuous
        interp_option=irrfao:average_gcell(0.0)+average_4pt+nearest_neighbor
        masked = water
        fill_missing = 0.
        rel_path=irrfao:irrigation/fao/
===============================

**Comment by barlage** [comment]
I have place the data here:

/glade/scratch/barlage/geogrid/geog

I suggest the nested directory structure since in our irrigation work we have multiple datasets and it is cleaner to have them collocated in an irrigation/ directory. My GEOGRID.TBL has the following entry:


===============================
name=IRRIGATION
        priority=1
        optional=yes
        dest_type=continuous
        interp_option=irrfao:average_gcell(0.0)+average_4pt+nearest_neighbor
        masked = water
        fill_missing = 0.
        rel_path=irrfao:irrigation/fao/
===============================

**Comment by barlage** [comment]
I have place the data here:

/glade/scratch/barlage/geogrid/geog

I suggest the nested directory structure since in our irrigation work we have multiple datasets and it is cleaner to have them collocated in an irrigation/ directory. My GEOGRID.TBL has the following entry:


===============================
name=IRRIGATION
        priority=1
        optional=yes
        dest_type=continuous
        interp_option=irrfao:average_gcell(0.0)+average_4pt+nearest_neighbor
        masked = water
        fill_missing = 0.
        rel_path=irrfao:irrigation/fao/
===============================

**Comment by barlage** [comment]
FYI, this is global 1/12 degree data for fraction of grid that is irrigated. Data source is FAO:

http://www.fao.org/aquastat/en/geospatial-information/global-maps-irrigated-areas

**Comment by dudhia** [comment]
> @arjanna @dudhia Can you verify that the code before and after this addition gives the same results with Noah? If it has been tested, please add it to the PR.
@arjanna We are still waiting for a reply to this question.

**Comment by davegill** [comment]
@weiwangncar @dudhia @arjanna 
Folks,
With my latest commit, this runs fine with ARW.
Fortran does not sequentially resolve logical statements, so we can't have
```
   if(present(x) .and. x >0 ) then
```
We have to split this into two pieces
```
   if(present(x)) then
      if(x > 0 ) then
```

**Comment by davegill** [comment]
Without my mod, this code failed a simple run with GNU. With the mod:
```
WRF NUMBER OF TILES =   1
Timing for main: time 2000-01-24_12:03:00 on domain   1:   12.10688 elapsed seconds
Timing for main: time 2000-01-24_12:06:00 on domain   1:    3.81167 elapsed seconds
Timing for main: time 2000-01-24_12:09:00 on domain   1:    3.58310 elapsed seconds
Timing for main: time 2000-01-24_12:12:00 on domain   1:    3.76686 elapsed seconds
Timing for main: time 2000-01-24_12:15:00 on domain   1:    3.60564 elapsed seconds
Timing for main: time 2000-01-24_12:18:00 on domain   1:    3.80850 elapsed seconds
Timing for main: time 2000-01-24_12:21:00 on domain   1:    3.62672 elapsed seconds
Timing for main: time 2000-01-24_12:24:00 on domain   1:   12.47272 elapsed seconds
Timing for main: time 2000-01-24_12:27:00 on domain   1:    3.64591 elapsed seconds
Timing for main: time 2000-01-24_12:30:00 on domain   1:    3.81491 elapsed seconds
Timing for Writing wrfout_d01_2000-01-24_12:30:00 for domain        1:    0.16084 elapsed seconds
d01 2000-01-24_12:30:00 wrf: SUCCESS COMPLETE WRF
```

**Pull request merged**